lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,594 @@
+    """Charm Event triggered when a TLS certificate is required."""
+
+    def __init__(self, handle, certificate_signing_request: str, relation_id: int):
+        super().__init__(handle)
+        self.certificate_signing_request = certificate_signing_request
+        self.relation_id = relation_id
+
+    def snapshot(self) -> dict:
+        """Returns snapshot."""
+        return {
+            "certificate_signing_request": self.certificate_signing_request,
+            "relation_id": self.relation_id,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restores snapshot."""
+        self.certificate_signing_request = snapshot["certificate_signing_request"]
+        self.relation_id = snapshot["relation_id"]
+
+
+def _load_relation_data(raw_relation_data: dict) -> dict:
+    """Loads relation data from the relation data bag.
+
+    Json loads all data.
+
+    Args:
+        raw_relation_data: Relation data from the databag
+
+    Returns:
+        dict: Relation data in dict format.
+    """
+    certificate_data = dict()
+    for key in raw_relation_data:
+        try:
+            certificate_data[key] = json.loads(raw_relation_data[key])
+        except json.decoder.JSONDecodeError:
+            certificate_data[key] = raw_relation_data[key]
+    return certificate_data
+
+
+def generate_private_key(
+    password: bytes,
+    key_size: int = 2048,
+    public_exponent: int = 65537,
+) -> bytes:
+    """Generates a private key.
+
+    Args:
+        password (bytes): Password for decrypting the private key
+        key_size (int): Key size in bytes
+        public_exponent: Public exponent.
+
+    Returns:
+        bytes: Private Key
+    """
+    private_key = rsa.generate_private_key(
+        public_exponent=public_exponent,
+        key_size=key_size,
+    )
+    key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.BestAvailableEncryption(password),
+    )
+    return key_bytes
+
+
+def generate_csr(
+    private_key: bytes, private_key_password: bytes, subject: str, sans: Optional[List[str]]
+) -> bytes:
+    """Generates a CSR using private key and subject.
+
+    Args:
+        private_key (bytes): Private key
+        private_key_password (bytes): Private key password
+        subject (str): CSR Subject.
+        sans (list): List of subject alternative names
+
+    Returns:
+        bytes: CSR
+    """
+    signing_key = serialization.load_pem_private_key(private_key, password=private_key_password)
+    csr = x509.CertificateSigningRequestBuilder(
+        subject_name=x509.Name(
+            [
+                x509.NameAttribute(x509.NameOID.COMMON_NAME, subject),
+            ]
+        )
+    )
+    if sans:
+        csr.add_extension(
+            x509.SubjectAlternativeName([x509.DNSName(san) for san in sans]), critical=False
+        )
```

Comment:
I think we should allow the extensions.

e.g. iif the application needs strange key usages (or other extensions), the app should generate its own `cryptography.x509.KeyUsage` object and pass it as a List into this function
```
   for extension in criticalExtensions:
       csr.add_extension(extension, critical=True)
```

Reply:
Done